rclcpp/include/rclcpp/rate.hpp

```diff
@@ -82,38 +96,61 @@ class GenericRate : public RateBase
       // Either way do not sleep and return false
       return false;
     }
+    // Calculate the time to sleep
+    auto time_to_sleep = next_interval - now;
     // Sleep (will get interrupted by ctrl-c, may not sleep full time)
-    rclcpp::sleep_for(time_to_sleep);
+    clock_->sleep_for(time_to_sleep);
```

Comment:
The `rclcpp::Clock::sleep_for()` - is a [wrapper](https://github.com/ros2/rclcpp/blob/20e9cd17f6723337dc0915f4e2a966f9585c309c/rclcpp/src/rclcpp/clock.cpp#L179-L183) for `rclcpp::Clock::sleep_until()`. `sleep_until()` seems to have a time [clock handler](https://github.com/ros2/rclcpp/blob/20e9cd17f6723337dc0915f4e2a966f9585c309c/rclcpp/src/rclcpp/clock.cpp#L140), catching time jumps and clock source changes. It seems, that during this change waiting is being interrupted [[1](https://github.com/ros2/rclcpp/blob/20e9cd17f6723337dc0915f4e2a966f9585c309c/rclcpp/src/rclcpp/clock.cpp#L146)], [[2](https://github.com/ros2/rclcpp/blob/20e9cd17f6723337dc0915f4e2a966f9585c309c/rclcpp/src/rclcpp/clock.cpp#L167)].

So, we have two types of situations:
1. Time was jumped (backward or forward) during the waiting:
  - In this case this situation will be handled correctly by the `rclcpp::Clock::sleep_until()`.
  One thing it looks I've missed: `sleep()` routine needs return the value from `sleep_for()` instead of always `true`:
  ```
  - clock_->sleep_for(time_to_sleep);
  - return true;
  + return clock_->sleep_for(time_to_sleep);
  ```
2. Time was jumped between `sleep()` beginning (when we already got `now` variable) and `clock_->sleep_for()` call. Very rare situation, but anyway:
 - In this case, inside `sleep_for()` the time to sleep will be [calculated](https://github.com/ros2/rclcpp/blob/20e9cd17f6723337dc0915f4e2a966f9585c309c/rclcpp/src/rclcpp/clock.cpp#L182) in already new time after jump, and we will sleep a required `time_to_sleep`, but in new time. Not fully correct situaion, but later on, on next `sleep()` call, if time was jumped (back or forward) more than on one `period_`, it will be handled inside the `sleep()`: for [back](https://github.com/AlexeyMerzlyakov/rclcpp/blob/7323e9730a8f3e288177bb30b74777f34e33dfee/rclcpp/include/rclcpp/rate.hpp#L82) jump and for [forward](https://github.com/AlexeyMerzlyakov/rclcpp/blob/7323e9730a8f3e288177bb30b74777f34e33dfee/rclcpp/include/rclcpp/rate.hpp#L89) jump; and situation will be stabilized.

Reply:
> I'll do it in next PR update.

Done in the [latest commit](https://github.com/ros2/rclcpp/pull/2123/commits/30b95de7cc242e1981091f406fb9c8226d34f97f#diff-c6ecfbeee7b7384c9e72a8984cc2fc718415b3bb741c0ea41ffc1d9b0718fea7R164)